packages/core/src/common/progress-service.ts

```diff
@@ -0,0 +1,81 @@
+/********************************************************************************
+ * Copyright (C) 2019 TypeFox and others.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License v. 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ *
+ * This Source Code may also be made available under the following Secondary
+ * Licenses when the conditions for such availability set forth in the Eclipse
+ * Public License v. 2.0 are satisfied: GNU General Public License, version 2
+ * with the GNU Classpath Exception which is available at
+ * https://www.gnu.org/software/classpath/license.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ ********************************************************************************/
+
+import { injectable, inject } from 'inversify';
+import {
+    Progress,
+    ProgressUpdate,
+    ProgressMessage
+} from './message-service-protocol';
+import { CancellationTokenSource } from './cancellation';
+import { ProgressClient } from './progress-service-protocol';
+import { MessageService } from './message-service';
+
+@injectable()
+export class ProgressService {
+
+    @inject(ProgressClient) protected readonly client: ProgressClient;
+    @inject(MessageService) protected readonly messageService: MessageService;
+
+    async showProgress(message: ProgressMessage, onDidCancel?: () => void): Promise<Progress> {
+        if (this.shouldDelegate(message)) {
```

Comment:
@akosyakov as discussed briefly, I'm going to add another change to deprecate `MessageService.showProgress`. In order to do so, I'm currently working on cancellation of notifications in general, which should allow to delegate to the `MessageService` here as well as reduce redundancy in code.